### PR TITLE
parallel implementation for all_pairs_bellman_ford_path

### PR DESCRIPTION
**Pull request by Schefflera-Arboricola**
https://github.com/networkx/networkx/issues/7003

Also, for a larger number of nodes, I was getting this heatmap : 
Screenshot 2023-10-11 at 9 51 46 AM

Please give your feedback.

Thank you :)

### CODE REVIEW

**Review by dschult — COMMENTED** [review]
The helper function could be simplified (and maybe sped up) by using a dict comprehension:
```python
def _calculate_shortest_paths_subset(G, chunk, weight):
    return {n: single_source_bellman_ford_path(n, weight=weight) for n in chunk}
```
and it might also be easy (since it is so short) to define the helper function inside the function itself. It's a style choice.  If there is a chance that it could be used by another function, then keep it defined in the main module.

I have a feeling that some of these idioms will be used over and over. Like:
```python
    num_in_chunk = max(len(nodes) // total_cores, 1)
    node_chunks = nxp.chunks(nodes, num_in_chunk)
```
PR #7 has consolidated some of those into utility functions that most of the functions in nx_parallel could use. There already is one utility function you have used: `nxp.cpu_count()`. Maybe there should be others. I'm not sure what the best way to implement things like that is. But we should think about it. No need to implement those ideas for this PR though. 

:)

**Comment by Schefflera-Arboricola** [comment]
@dschult There wasn't much difference from the above dict comprehension also. The parallel implementation is taking more time because here I am computing all the paths and returning them in a dictionary but in the [non-parallel implementation](https://github.com/networkx/networkx/blob/a40863040eb31f60cf39a1c7e4255e059007c461/networkx/algorithms/shortest_paths/weighted.py#L1932C11-L1932C11) an iterator is returned so the paths are yielded(computed) when/if the iterator is iterated later in the program. And we cannot break down the task, such that, we compute multiple generators in parallel and then combine them, like this : 
```.py 
def _calculate_shortest_paths_subset(G, chunk, weight):
    for n in chunk:
        yield n, single_source_bellman_ford_path(G, n, weight=weight)
```  
because `joblib` requires a serializable data structure to be returned to apply parallization(error: `TypeError: cannot pickle 'generator' object`). Probably because generators are stateful, so if would try to access a generator through multiple processes at the same time, then it will probably give incorrect outputs. Also, it doesn't make much sense to parallelly generate generators. I couldn’t find a way around it. 

But if someone definitely wants to compute all the paths then this implementation is better than iterating through the iterator of the non-parallel function. Here, the time to convert the iterator into a dictionary is taken into account ([timing file](https://github.com/networkx/nx-parallel/pull/14/files#diff-59a12e50776b5cf7c07dabaf6ca8c18ce9f7598d9815619959fb2e563209b725R26))
```diff
            t1 = time.time()
            c = currFun(H)
+           d1=dict(c)
            t2 = time.time()
            parallelTime = t2 - t1
            t1 = time.time()
            c = currFun(G)
+           d2=dict(c)
            t2 = time.time()
            stdTime = t2 - t1
            timesFaster = stdTime / parallelTime
            heatmapDF.at[j, i] = timesFaster
            print("Finished " + str(currFun))
```
here, we are getting speedups : 
Screenshot 2023-10-16 at 11 54 24 AM

Also, thanks for the feedback on styling the code!
Please let me know what you think and how I should proceed from here.

Thank you :)

**Comment by dschult** [comment]
I think you can handle generators with joblib.
Take a look at this page and maybe a page linked from this one.
https://joblib.readthedocs.io/en/stable/parallel.html
Hopefully it is simple to do.

my understanding is that each cpu generates values, but they are yielded to the user in the order they are started, so if one is faster than a previous one, it waits until the previous one is finished and yielded before being yielded itself. The order is preserved relative to the non-parallel version.

now, the timing should still include generating the entire set of nodes. Because setting up the generators take very little time. It is when they actually do the computations that it takes time.

**Comment by Schefflera-Arboricola** [comment]
thanks @dschult 

I have updated it now. Let me know if there's anything else to improve upon.

Thank you :)

**Review by dschult — COMMENTED** [review]
This looks good -- I have some comments below, but I think this is close to being ready.
:)

File: nx_parallel/algorithms/shortest_paths/weighted.py [hunk 1]
```diff
@@ -30,40 +30,32 @@ def all_pairs_bellman_ford_path(G, weight="weight"):
...
+        shortest path as the key value.
 
     Notes
     -----
     Edge weight attributes must be numerical.
     Distances are calculated as sums of weighted edges traversed.
 
     """
+    def _calculate_shortest_paths_subset(G, source, weight):
+        return (source, single_source_bellman_ford_path(G, source, weight=weight))
```
[the remark this answers is not in the document]
Comment (Schefflera-Arboricola) [reply nx_parallel/algorithms/shortest_paths/weighted.py:44]: yes, passing only `source` is a little faster, following are the speed-ups for the same :  

Screenshot 2023-11-03 at 6 04 59 PM

File: timing/timing_individual_function.py [hunk 2]
```diff
@@ -11,68 +13,74 @@
...
-            H = nx_parallel.ParallelGraph(G)
+        # time both versions and update heatmapDF
+        t1 = time.time()
+        c = currFun(H)
+        if type(c)==types.GeneratorType: d = dict(c) 
+        t2 = time.time()
+        parallelTime = t2 - t1
+        t1 = time.time()
+        c = currFun(G)
+        if type(c)==types.GeneratorType: d = dict(c)
```
[the remark this answers is not in the document]
Comment (Schefflera-Arboricola) [reply timing/timing_individual_function.py:39]: I checked `d1==d2` for the all_pairs_bellman_ford_path before committing. It was true for all cases. But, for betweenness_centrality it was not always true. I had to round up all the values. I can add separate tests for all the algorithms, if that seems good to you.

**Comment by Schefflera-Arboricola** [comment]
@dschult I have made all the updates please let me know if this looks good to you. Thank you very much for the review :)

**Comment by Schefflera-Arboricola** [comment]
I tried running the same timing_individual_function.py script again and I got the following heatmap : 
Screenshot 2023-11-04 at 12 37 35 PM
This is the heatmap right now in the PR : 
Screenshot 2023-11-04 at 12 39 53 PM

Would you say it's a lot of variation, especially for the 100-node graphs?

Thank you :)

**Review by dschult — APPROVED** [review]
I think this looks good. 
There are some big-picture issues we should probably consider at some point -- but not for this PR's intent (addition of a parallel implementation). 

Some of the questions invovle things like:
- should we include the NX docs information in the nx-parallel doc_strings? That might lead to one version becoming out-of-date compared to the other. Perhaps we should only put info about the parallel implementation here and refer to the NX docs for the info about the original function.  If so, where do we draw the line? at the function signature? do we copy the parameter descriptions?
- Should we centralize the 'chunk'ing and 'map'ing and 'reduce'ing?  see #7 

But let's go ahead and merge this and worry about the big picture issues in another PR.
Thanks!

File: nx_parallel/algorithms/shortest_paths/weighted.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    def _calculate_shortest_paths_subset(source):
+        return (source, single_source_bellman_ford_path(G, source, weight=weight))
+
+    if hasattr(G, "graph_object"):
+        G = G.graph_object
+
+    nodes = G.nodes
+
+    total_cores = nxp.cpu_count()
```
Comment (MridulS) [inline nx_parallel/algorithms/shortest_paths/weighted.py:66]: We should be moving this to the function signature and match something like what scikit-learn does (`n_jobs`) but not a blocker here.

**Comment by MridulS** [comment]
Thanks for this @Schefflera-Arboricola !

There are a bunch of stuff we should revisit, especially in the context of #7 but for now let's merge this in :)

**Pull request merged**